rmw_connext_shared_cpp/src/init.cpp

```diff
@@ -12,23 +12,242 @@
+      factory_qos.resource_limits.max_objects_per_thread = 8192;
+      dpf->set_qos(factory_qos);
+
+      switch (set_default_qos_library(dpf)) {
+        case TristateRetCode::SET:
+          if (!set_default_qos_profile(dpf)) {
+            ret = RMW_RET_ERROR;
+          }
+          break;
+        case TristateRetCode::NOT_SET:
+          break;
+        default:  // fallthrough
+        case TristateRetCode::FAILED:
+          ret = RMW_RET_ERROR;
+          return;
+      }
+
+      switch (is_env_variable_set("RMW_CONNEXT_IGNORE_ROS_QOS")) {
+        case TristateRetCode::SET:
+          g_is_ros_qos_ignored = true;
+          break;
+        case TristateRetCode::NOT_SET:
+          break;
+        default:  // fallthrough
+        case TristateRetCode::FAILED:
+          ret = RMW_RET_ERROR;
+          return;
+      }
+
+      switch (is_env_variable_set("RMW_CONNEXT_ALLOW_NODE_QOS_PROFILES")) {
+        case TristateRetCode::SET:
+          g_are_node_profiles_allowed = true;
+          break;
+        case TristateRetCode::NOT_SET:
+          break;
+        default:  // fallthrough
+        case TristateRetCode::FAILED:
+          ret = RMW_RET_ERROR;
+          return;
+      }
+      switch (is_env_variable_set("RMW_CONNEXT_DO_NOT_OVERRIDE_PUBLICATION_MODE")) {
+        case TristateRetCode::SET:
+          g_is_publish_mode_overriden = false;
+          break;
+        case TristateRetCode::NOT_SET:
+          break;
+        default:  // fallthrough
+        case TristateRetCode::FAILED:
+          ret = RMW_RET_ERROR;
+          return;
+      }
+    }
+  );
+  return ret;
+}
+
+static TristateRetCode
+set_default_qos_library(DDS::DomainParticipantFactory * dpf)
+{
+  const char * qos_profile_library_name = NULL;
+  const char * error = rcutils_get_env(
+    "RMW_CONNEXT_QOS_PROFILE_LIBRARY", &qos_profile_library_name);
+  if (error) {
+    RMW_SET_ERROR_MSG_WITH_FORMAT_STRING("rcutils_get_env() failed: '%s'", error);
+    return TristateRetCode::FAILED;
+  }
+  if (qos_profile_library_name && 0 == strcmp("", qos_profile_library_name)) {
+    qos_profile_library_name = NULL;
+  }
+
+  DDS_StringSeq qos_libraries;
+  if (!qos_profile_library_name) {
+    // environment variable is empty
+    if (DDS::RETCODE_OK != dpf->get_qos_profile_libraries(qos_libraries)) {
+      RMW_SET_ERROR_MSG("failed to get qos profile libraries");
+      return TristateRetCode::FAILED;
+    }
+    // If only one non-builtin qos profile library was loaded, use that one.
+    if (qos_libraries.length() > 3) {
+      return TristateRetCode::NOT_SET;
+    }
+    for (int i = 0; i < qos_libraries.length(); i++) {
+      if (
+        strcmp("BuiltinQosLib", qos_libraries[i]) != 0 &&
+        strcmp("BuiltinQosLibExp", qos_libraries[i]) != 0)
+      {
+        if (!qos_profile_library_name) {
+          qos_profile_library_name = qos_libraries[i];
+        } else {
+          qos_profile_library_name = NULL;
+          break;
```

Comment:
This logic looks odd to me. IIUC, we're only setting the library name if there is exactly one was loaded that is not one of the two built-ins (and there are only ever two built-ins?). Is that right? If so, maybe adding a comment here would help clarify:
```suggestion
          // More than one QoS profile library was loaded, do not set
          qos_profile_library_name = NULL;
          break;
```

Reply:
> and there are only ever two built-ins?

yeah, https://community.rti.com/examples/built-qos-profiles.

> Is that right? If so, maybe adding a comment here would help clarify:

Sounds good